wallet-new/src/Cardano/Wallet/Kernel/DB/AcidState.hs

```diff
@@ -366,13 +378,91 @@ updateHdAccountName accId name = do
     a <- runUpdate' . zoom dbHdWallets $ HD.updateHdAccountName accId name
     get >>= \st' -> return $ bimap identity (st',) a
 
-deleteHdRoot :: HdRootId -> Update DB ()
-deleteHdRoot rootId = runUpdateNoErrors . zoom dbHdWallets $
+deleteHdRoot :: HdRootId -> Update DB (Either UnknownHdRoot ())
+deleteHdRoot rootId = runUpdate' . zoom dbHdWallets $
     HD.deleteHdRoot rootId
 
+-- | Deletes the 'HdAccount' identified by the input 'HdAccountId' together
+-- with all the linked addresses.
 deleteHdAccount :: HdAccountId -> Update DB (Either UnknownHdAccount ())
-deleteHdAccount accId = runUpdate' . zoom dbHdWallets $
-    HD.deleteHdAccount accId
+deleteHdAccount accId = do
+    db <- get
+    case HD.readAddressesByAccountId accId (db ^. dbHdWallets) of
+         Left accNotFound -> return (Left accNotFound)
+         Right allAddresses -> do
+             -- Deletes all the children addresses.
+             res <- foldM delete (Right ()) allAddresses
```

Comment:
Let's just have a `runUpdate'` at the top, so that we can use `mapM` here. `runUpdate'` is basically only used top-level; the purpose of `Update'` is to provide proper error threading so that these kinds of shenanigans are not necessary :)

Reply:
All this indent-indent-indent won't be necessary anymore either

Reply:
> anymore either

See what you did here, mr good puns :trollface:

Reply:
_groan_